In [1]:
pip install numpy pandas sklearn

  Preparing metadata (setup.py) ... done
  Created wheel for sklearn: filename=sklearn-0.0.post1-py3-none-any.whl size=2936 sha256=0ea25b87e3ea7854867286d94a76d6315dc38acdd9496e7da134ee8ff5662815
  Stored in directory: /Users/saan/Library/Caches/pip/wheels/f8/e0/3d/9d0c2020c44a519b9f02ab4fa6d2a4a996c98d79ab2f569fa1
Successfully built sklearn
Note: you may need to restart the kernel to use updated packages.


In [2]:
# Importing libraries
import numpy as np
import pandas as pd
import itertools
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import PassiveAggressiveClassifier
from sklearn.metrics import accuracy_score, confusion_matrix

In [3]:
# Reading data from csv file
df = pd.read_csv('/Users/saan/Downloads/news.csv')

In [5]:
# What does the data look like?
df.shape
df.head()

Unnamed: 0                                              title  \
0        8476                       You Can Smell Hillary’s Fear   
1       10294  Watch The Exact Moment Paul Ryan Committed Pol...   
2        3608        Kerry to go to Paris in gesture of sympathy   
3       10142  Bernie supporters on Twitter erupt in anger ag...   
4         875   The Battle of New York: Why This Primary Matters   

                                                text label  
0  Daniel Greenfield, a Shillman Journalism Fello...  FAKE  
1  Google Pinterest Digg Linkedin Reddit Stumbleu...  FAKE  
2  U.S. Secretary of State John F. Kerry said Mon...  REAL  
3  — Kaydee King (@KaydeeKing) November 9, 2016 T...  FAKE  
4  It's primary day in New York and front-runners...  REAL

In [6]:
# Getting the labels - Fake/Real
labels = df.label
labels.head()

0    FAKE
1    FAKE
2    REAL
3    FAKE
4    REAL
Name: label, dtype: object

In [7]:
# Splitting the dataset
x_train, x_test, y_train, y_test = train_test_split(df['text'], labels, test_size = 0.2, random_state = 7)

In [10]:
# Initializing TfidfVectorizer - removing stop words and setting max document frequency
vectorizer = TfidfVectorizer(stop_words = 'english', max_df = 0.7)

# Fit and transform train set
train = vectorizer.fit_transform(x_train)

# Transform test set
test = vectorizer.transform(x_test)

In [12]:
# Initializing a PassiveAggressiveClassifier
pac = PassiveAggressiveClassifier(max_iter = 50)
pac.fit(train, y_train)

# Predicting test set
y_pred = pac.predict(test)

# Calculating accuracy of the model
score = accuracy_score(y_test, y_pred)

print(f'Accuracy: {round(score * 100, 2)}')

Accuracy: 92.5


In [14]:
# Confusion matrix
confusion_matrix(y_test, y_pred, labels=['FAKE','REAL'])


array([[588,  50],
       [ 45, 584]])

Result:
Through this model of accuracy 92.5%
we have:
    588 true negative
    584 true positive
    45 false positive
    50 false negative